File: src/GetIt/Models/Costume.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace GetIt
{
    /// <summary>
    /// A player outfit.
    /// </summary>
    public class Costume
    {
        /// <summary>
        /// Size of the costume.
        /// </summary>
        public Size Size { get; }
        
        /// <summary>
        /// List of paths that define the costume.
        /// </summary>
        public IImmutableList<GeometryPath> Paths { get; }

        /// <summary>
        /// Initializes all properties of an instance.
        /// </summary>
        public Costume(Size size, IImmutableList<GeometryPath> paths)
        {
            Size = size;
            Paths = paths;
        }

        /// <summary>
        /// Creates a costume that has the shape of a polygon.
        /// </summary>
        /// <param name="fillColor">The color that is used to fill the polygon.</param>
        /// <param name="points">The points that define the polygon.</param>
        /// <returns>The created costume.</returns>
        public static Costume CreatePolygon(RGBA fillColor, IEnumerable<Position> points)
        {
            points = points
                .Select(p => new Position(p.X, -p.Y))
                .ToList();

            var minX = points.Min(p => p.X);
            var maxX = points.Max(p => p.X);
            var minY = points.Min(p => p.Y);
            var maxY = points.Max(p => p.Y);

            var transformedPoints = points
                .Select(p => new Position(p.X - minX, p.Y - minY))
                .Select((p, i) => $"{(i == 0 ? "M" : "L")} {p.X},{p.Y}");
            var path = $"{string.Join(" ", transformedPoints)} Z";

            return new Costume(
                new Size(maxX - minX, maxY - minY),
                ImmutableList<GeometryPath>.Empty
                    .Add(new GeometryPath(fillColor, path))
            );
        }

        /// <summary>
        /// Creates a costume that has the shape of a polygon.
        /// </summary>
        /// <param name="fillColor">The color that is used to fill the polygon.</param>
        /// <param name="points">The points that define the polygon.</param>
        /// <returns>The created costume.</returns>
        public static Costume CreatePolygon(RGBA fillColor, params Position[] points)
        {
            return CreatePolygon(fillColor, points.AsEnumerable());
        }

        /// <summary>
        /// Creates a costume that has the shape of a rectangle.
        /// The left bottom corner is at (0,0).
        /// </summary>
        /// <param name="fillColor">The color that is used to fill the rectangle.</param>
        /// <param name="size">The size of the rectangle.</param>
        /// <returns>The created costume.</returns>
        public static Costume CreateRectangle(RGBA fillColor, Size size)
        {
            return CreatePolygon(
                fillColor,
                new Position(0, 0),
                new Position(0, size.Height),
                new Position(size.Width, size.Height),
                new Position(size.Width, 0));
        }

        /// <summary>
        /// Creates a costume that has the shape of a circle.
        /// The center point is at (radius,radius).
        /// </summary>
        /// <param name="fillColor">The color that is used to fill the circle.</param>
        /// <param name="radius">The radius of the circle.</param>
        /// <returns>The created costume.</returns>
        public static Costume CreateCircle(RGBA fillColor, double radius)
        {
            return new Costume(
                new Size(2 * radius, 2 * radius),
                ImmutableList<GeometryPath>.Empty
                    .Add(new GeometryPath(fillColor, $"M 0,{radius} A {radius},{radius} 0 1 0 {2 * radius},{radius} A {radius},{radius} 0 1 0 0,{radius}"))
            );
        }
    }
}

File: GetIt.Common/Costume.fs
namespace GetIt

open System

/// Defines a player costume.
type Costume =
    {
        /// The size of the costume.
        Size: Size
        /// The definition of the costume in SVG format.
        SvgData: string
    }
    with
        ///<summary>Creates a circle costume.</summary>
        ///<param name="fillColor">The color that is used to fill the circle.</param>
        ///<param name="radius">The radius of the circle.</param>
        static member CreateCircle (fillColor, radius) =
            let size = { Width = 2. * radius; Height = 2. * radius }
            {
                Size = size
                SvgData =
                    sprintf """<svg width="%f" height="%f"><circle cx="%f" cy="%f" r="%f" style="fill:%s;fill-opacity:%f" /></svg>"""
                        size.Width size.Height radius radius radius (RGBAColor.rgbHexNotation fillColor) (RGBAColor.transparency fillColor)
            }

        ///<summary>Creates a polygon costume.</summary>
        ///<param name="fillColor">The color that is used to fill the polygon.</param>
        ///<param name="points">The points that define the polygon.</param>
        static member CreatePolygon (fillColor, [<ParamArray>] points) =
            let points = Array.toList points

            let xs = points |> List.map (fun p -> p.X)
            let ys = points |> List.map (fun p -> p.Y)
            let minX = List.min xs
            let maxX = List.max xs
            let minY = List.min ys
            let maxY = List.max ys

            let pointString =
                points
                |> List.map (fun p -> sprintf "%f,%f" p.X (maxY - p.Y))
                |> String.concat " "

            let size = { Width = maxX - minX; Height = maxY - minY }
            { Size = size
              SvgData =
                sprintf """<svg width="%f" height="%f"><polygon points="%s" style="fill:%s;fill-opacity:%f;" /></svg>"""
                    size.Width size.Height pointString (RGBAColor.rgbHexNotation fillColor) (RGBAColor.transparency fillColor)
            }

        ///<summary>Creates a rectangle costume.</summary>
        ///<param name="fillColor">The color that is used to fill the rectangle.</param>
        ///<param name="size">The size of the rectangle.</param>
        static member CreateRectangle (fillColor, size) =
            let points =
                [|
                    { X = 0.; Y = 0. }
                    { X = 0.; Y = size.Height }
                    { X = size.Width; Y = size.Height }
                    { X = size.Width; Y = 0. }
                |]
            Costume.CreatePolygon (fillColor, points)
